astropy/io/fits/tests/test_connect.py

```diff
@@ -304,6 +305,42 @@ def check_equal(filename, expected, start_from=1):
         check_equal(filename, 3, start_from=2)
         assert equal_data(t2, Table.read(filename, hdu=1))
 
+    def test_inexact_format_parse_on_read(self, tmpdir):
+        filename = str(tmpdir.join('test_inexact_format_parse_on_read.fits'))
+        c1 = fits.Column(name='a', array=np.array([1, 2, np.nan]), format='E')
+        table_hdu = fits.TableHDU.from_columns([c1])
+        table_hdu.writeto(filename)
+
+        with fits.open(filename) as hdulist:
+            assert read_table_fits(hdulist)
+
+            data = hdulist[1].data
+
+            for col in data.columns:
+                with pytest.raises(TypeError):
+                    assert np.isnan(col.array)
+
+    def test_null_format_parse_on_read(self, tmpdir):
+        filename = str(tmpdir.join('test_null_format_parse_on_read.fits'))
+        col = fits.Column(name='a', array=np.array([1, 2, 99, 60000], dtype='u2'), format='I', null=99, bzero=32768)
+        bin_table_hdu = fits.BinTableHDU.from_columns([col])
+        bin_table_hdu.writeto(filename, overwrite=True)
+
+        with fits.open(filename) as hdulist:
+            data = hdulist[1].data
+
+            for col in data.columns:
+                # The dtype scaling when using col.array fails for null definitions
+                mask = col.array == col.null
+                assert not all(mask)
+
+                # Accessing through data[col.name] succeeds with parsing null values
+                mask = data[col.name] == col.null
+                assert mask[2]
+
+        tab = Table.read(filename)
+        assert tab.mask[2]
```

Comment:
Same here, `col.array`, `col.null` etc. are tested on other files, so something like `Table.read(...)['col0'].mask == [False, False, True, False]` should be enough.

Reply:
That's what this effectively does: `tab.mask` returns the table's mask (1d array in this case) whose third element is `True` (the assert).

`['col0'].mask == [False, False, True, False]` won't work since you can't compare lists.

Reply:
Yes, my comment was mostly for the block above using the io.fits API. 
Also pytest would compare lists, but since it's a numpy array on the left you would need to wrap this in `np.all(...)`. For simple test case like this I find it useful to check all elements just to be sure that only the expected one is masked, but your `assert tab.mask[2]` would work too :+1:

Reply:
Good point. Are the changes sufficient?